### PR TITLE
plugin/file: Rename do to walk and cleanup and document

### PR DESCRIPTION
**Pull request by miekg**
* This renames Do to Walk to be more inline with Go standards.
* It adds Types to TypesForWildcard to show the different functions
* *Removes* the identical RR check when inserting; this was only done
  for A, AAAA and MX and not finished; removed under the mantra garbage
  in garbage out.

### CODE REVIEW

**Comment by miekg** [comment]
See #2986

**Comment by miekg** [comment]
this would  be good to merge as a simplifies some bits. It helps me for *sign* and I also have some ideas to fix the reload deadlock

File: plugin/file/tree/walk.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package tree
+
+import "github.com/miekg/dns"
+
+// Walk performs fn on all values stored in the tree. If a non-nill error is returned the
```
Comment (stp-ip) [inline plugin/file/tree/walk.go]:
```suggestion
// Walk performs fn on all values stored in the tree. If a non-nil error is returned the
```

File: plugin/file/tree/tree.go [hunk 2]
```diff
@@ -286,12 +286,9 @@ func (t *Tree) Delete(rr dns.RR) {
...
 		return
 	}
-	// Delete from this element.
-	empty := el.Delete(rr)
-	if empty {
-		t.deleteNode(rr)
-		return
-	}
+	el.Delete(rr)
+	t.deleteNode(rr)
```
Comment (stp-ip) [inline plugin/file/tree/tree.go]: Why delete the node either way?

**Review by stp-ip — COMMENTED** [review]
Still working through the code. I agree it makes it a bit cleaner, but not sure I am comfortable giving a lgtm yet.

File: plugin/file/tree/tree.go (hunk 2 above)
Comment (miekg) [reply plugin/file/tree/tree.go]: good question. Let me also add some test for this, because we don't have many... Note delete/walk is currently unused.

Comment (miekg) [reply plugin/file/tree/tree.go]: Added test and change this bit. Thanks for catching this.

**Comment by miekg** [comment]
I intent to merge this soon. I also need another function which is hard to do externally. A `WalkAuthoritative` this skips authoritative bits in a zone (i.e. records under a delegation).

Once these bits are in the locking can be simplified to only have 1 mutex, for reload and transfer.

**Pull request merged**